rir/src/compiler/opt/elide_env_spec.cpp

```diff
@@ -16,42 +16,96 @@ void ElideEnvSpec::apply(RirCompiler&, ClosureVersion* function,
+                    nextBB = nextBB->trueBranch();
+                return nextInstr(*(nextBB->begin()), nextBB, false);
+            }
+            return i;
+        };
+
     Visitor::run(function->entry, [&](BB* bb) {
+        std::unordered_map<Value*, Value*> stubFor;
         auto ip = bb->begin();
         while (ip != bb->end()) {
             Instruction* i = *ip;
             auto next = ip + 1;
 
-            if (i->envOnlyForObj() && i->hasEnv()) {
-                bool objectExpected = false;
-
-                i->eachArg([&](Value* arg) {
-                    if (arg != i->env())
-                        if (arg->type.maybeObj() &&
-                            arg->typeFeedback.maybeObj())
-                            objectExpected = true;
-                });
-
-                if (!objectExpected) {
-                    if (auto cp = checkpoint.at(i)) {
-                        i->elideEnv();
-                        i->eachArg([&](Value* arg) {
-                            if (arg != i->env())
-                                if (arg->type.maybeObj()) {
-                                    auto condition = new IsObject(arg);
-                                    ip = bb->insert(ip, condition);
-                                    ip++;
-                                    ip = bb->insert(
-                                        ip, (new Assume(condition, cp))->Not());
-                                    ip++;
-                                }
-                        });
-                        next = ip + 1;
-                        i->type.setNotObject();
+            if (i->hasEnv()) {
+                // Speculatively elide environments on instructions in which
+                // both operators are primitive values
+                if (checkpoint.at(i) && i->envOnlyForObj() &&
+                    nonObjectArgs(i)) {
+                    i->elideEnv();
+                    i->eachArg([&](Value* arg) {
+                        if (arg != i->env())
+                            if (arg->type.maybeObj()) {
+                                auto condition = (new TypeTest(arg))->object();
+                                ip = bb->insert(ip, condition);
+                                ip++;
+                                ip = bb->insert(
+                                    ip,
+                                    (new Assume(condition, checkpoint.at(i)))
+                                        ->Not());
+                                ip++;
+                            }
+                    });
+                    next = ip + 1;
+                    i->type.setNotObject();
+                }
+
+                // Speculatively elide envs on forces that only require them in
+                // case they access promises reflectively
+                if (auto force = Force::Cast(i)) {
+                    auto nextInstruction = nextInstr(force, bb, true);
+                    if (checkpoint.at(nextInstruction)) {
```

Comment:
ad-hoc searching for deopt point, that seems to ignore many edge-cases. For example, how about:
```
  force
  branch a b
a:
  checkpoint
  ...
b:
  return
```

this would just put the assume in the a branch and ignore the b branch, not?

Reply:
Another example:
```
    force
    effect
    checkpoint
```
I think it would skip effect, not?

Reply:
I see the problem here. `AvailableCheckpoints` was written for eager checks, but has no `nextCheckpoint` api, required for lazy checks. This needs some thinking. Do we need a to use a backwards analysis to get the nextCheckpoint?

Reply:
Exactly, this was the first attempt to loop for a forward checkpoint when we have not had the analysis. After the analysis the problem remained there. 

For the examples you wrote, I do not know, my point is that the invariant is that after a force, for which we could speculate, we must have a checkpoint without having effects in between. Then, moving the checkpoint to the branch may happen and that breaks this code true. But why would we do that? 

Summary, yes, a backward analysis would be the proper solution and would simplify the code of this motion by removing the nextInstr function. But I am not sure how to do that soundly in case we can move checkpoints to branches.

Reply:
> But why would we do that?

The pir passes have to be independently correct. relying on the fact that other passes (currently) don't generate a certain kind of graph is not what we want. if we guarantee a certain invariant on the code, it has to be checked in the verifier!

how to do it soundly: exactly like the availableassumptions, but with a backwards analysis (you can plug and play forward/backward analysis into a static analysis implementation)